examples/src/test/java/io/grpc/examples/header/HeaderClientInterceptorTest.java

```diff
@@ -0,0 +1,133 @@
+import io.grpc.Server;
+import io.grpc.ServerCall;
+import io.grpc.ServerCall.Listener;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerInterceptor;
+import io.grpc.ServerInterceptors;
+import io.grpc.StatusRuntimeException;
+import io.grpc.examples.helloworld.GreeterGrpc;
+import io.grpc.examples.helloworld.GreeterGrpc.GreeterBlockingStub;
+import io.grpc.examples.helloworld.GreeterGrpc.GreeterImplBase;
+import io.grpc.examples.helloworld.HelloReply;
+import io.grpc.examples.helloworld.HelloRequest;
+import io.grpc.inprocess.InProcessChannelBuilder;
+import io.grpc.inprocess.InProcessServerBuilder;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Matchers;
+
+import java.io.IOException;
+
+/**
+ * Unit tests for {@link HeaderClientInterceptor}.
+ * For demonstrating how to write gRPC unit test only.
+ * Not intended to provide a high code coverage or to test every major usecase.
+ *
+ * <p>For basic unit test examples see {@link io.grpc.examples.helloworld.HelloWorldClientTest} and
+ * {@link io.grpc.examples.helloworld.HelloWorldServerTest}.
+ */
+@RunWith(JUnit4.class)
+public class HeaderClientInterceptorTest {
+
+  private final ServerInterceptor mockServerInterceptor =
+      spy(new ServerInterceptor() {
+        @Override
+        public <ReqT, RespT> Listener<ReqT> interceptCall(ServerCall<ReqT, RespT> call,
+            Metadata headers,
+            ServerCallHandler<ReqT, RespT> next) {
+          return next.startCall(call, headers);
+        }
+      });
+
+  private Server fakeServer;
+  private ManagedChannel inProcessChannel;
+
+  @Before
+  public void setUp() throws IOException {
+    String uniqueServerName = "fake server for " + getClass();
+    fakeServer = InProcessServerBuilder.forName(uniqueServerName)
+        .addService(ServerInterceptors.intercept(new GreeterImplBase() {}, mockServerInterceptor))
+        .build()
+        .start();
+    inProcessChannel = InProcessChannelBuilder.forName(uniqueServerName)
+        .intercept(new HeaderClientInterceptor())
+        .build();
+  }
+
+  @After
+  public void tearDown() {
+    if (inProcessChannel != null) {
+      inProcessChannel.shutdownNow();
+    }
+    if (fakeServer != null) {
+      fakeServer.shutdownNow();
+    }
+  }
+
+  @Test
+  public void clientHeaderDeliveredToServer() {
+    GreeterBlockingStub blockingStub = GreeterGrpc.newBlockingStub(inProcessChannel);
+    ArgumentCaptor<Metadata> metadataCaptor = ArgumentCaptor.forClass(Metadata.class);
+
+    try {
+      blockingStub.sayHello(HelloRequest.getDefaultInstance());
+      fail();
+    } catch (StatusRuntimeException expected) {
+      // expected because the method is not implemented at server side
+    }
+
+    verify(mockServerInterceptor)
+        .interceptCall(
+            Matchers.<ServerCall<HelloRequest, HelloReply>>any(),
```

Comment:
`isA(ServerCall.class)` + `@SuppressWarnings("unchecked")` is losing <Type> parameter information. It's equivalent to using `any()` here as the first argument of `interceptCall()` is always of `ServerCall` class.